tezos/encoding/src/binary_reader.rs

```diff
@@ -3,874 +3,34 @@
 
 //! Tezos binary data reader.
 
-use failure::ResultExt;
-use std::fmt::{self, Display};
+use std::fmt;
 
-use bit_vec::BitVec;
-use bytes::Buf;
 use failure::Fail;
-use serde::de::Error as SerdeError;
-
-use crate::bit_utils::{BitReverse, BitTrim, Bits, ToBytes};
-use crate::de;
-use crate::encoding::{Encoding, Field, SchemaType};
-use crate::types::{self, Value};
-
-use super::error_context::EncodingError;
 
 /// Error produced by a [BinaryReader].
-pub type BinaryReaderError = EncodingError<BinaryReaderErrorKind>;
-
-/// Actual size of encoded data
-#[derive(Debug, Clone, Copy)]
-pub enum ActualSize {
-    /// Exact size
-    Exact(usize),
-    /// Unknown size exceeding the limit
-    GreaterThan(usize),
-}
-
-impl Display for ActualSize {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            ActualSize::Exact(size) => write!(f, "{}", size),
-            ActualSize::GreaterThan(size) => write!(f, "greater than {}", size),
-        }
-    }
-}
-
-/// Kind of error for [BinaryReaderError]
-#[derive(Debug, Fail, Clone)]
-pub enum BinaryReaderErrorKind {
-    /// More bytes were expected than there were available in input buffer.
-    #[fail(display = "Input underflow, missing {} bytes", bytes)]
-    Underflow { bytes: usize },
-    /// Writer tried to write too many bytes to output buffer.
-    #[fail(display = "Input overflow, excess of {} bytes", bytes)]
-    Overflow { bytes: usize },
-    /// Generic deserialization error.
-    #[fail(display = "Deserializer error: {}", error)]
-    DeserializationError { error: crate::de::Error },
-    /// No tag with the corresponding id was found. This might not be an error of the binary data but
-    /// may simply mean that we have not yet defined tag in encoding.
-    #[fail(display = "No tag found for id: 0x{:X}", tag)]
-    UnsupportedTag { tag: u16 },
-    /// Encoding boundary constraint violation
-    #[fail(
-        display = "Encoded data {} exceeded its size boundary: {}, actual: {}",
-        name, boundary, actual
-    )]
-    EncodingBoundaryExceeded {
-        name: String,
-        boundary: usize,
-        actual: ActualSize,
-    },
-    /// Arithmetic overflow
-    #[fail(display = "Arithmetic overflow while encoding {:?}", encoding)]
-    ArithmeticOverflow { encoding: &'static str },
-}
+#[derive(Debug, Clone, Fail)]
+pub struct BinaryReaderError(String, bool);
```

Comment:
@akoptelov 
well, the last thing,
I am not sure about this BinaryReaderError,
it looks like there are two separeated things joined together:
- String for generic error message
- bool for is_unsupported_tag

you can create "generic message" + true, which does not make sense for scenario,

I would maybe do it clearly:
```
enum BinaryReaderError {
   GenericError(String)
   UnsupportedTag(<tag>)
}
```

or something like that

Reply:
Hmm... Actually it should be treated differently for `PeerResponseMessage` where we expect unsupported/unknown tags and can just ignore them, and e.g. for `Ack` where we have strict set of variants, and unexpected variant should be taken as a parsing failure... So I've added some code for that and changed the error as you suggested.